components/net_traits/lib.rs

```diff
@@ -203,7 +203,9 @@ pub trait FetchTaskTarget {
 pub trait FetchResponseListener {
     fn process_request_body(&mut self);
     fn process_request_eof(&mut self);
-    fn process_response(&mut self, metadata: Result<Metadata, NetworkError>);
+    fn process_response(&mut self,
+                        metadata: Result<Metadata, NetworkError>,
+                        actual_metadata: Option<Result<Metadata, NetworkError>>);
```

Comment:
I propose a different solution here, since I think this makes it really easy to ignore the implications of what response consumer code actually ends up interacting with. Instead, what about:
```rust
enum FetchResponse {
    Unfiltered(Metadata),
    Filtered(FilteredResponse),
}

pub struct FilteredResponse {
    filtered: Metadata,
    unsafe_: Metadata,
}

impl FilteredResponse {
    pub fn safe_response(self) -> Metadata {
        self.filtered
    }
    pub fn unsafe_response(self) -> Metadata {
        self.unsafe_
    }
}

trait FetchResponseListener {
    fn process_response(&mut self, metadata: Result<FetchResponse, NetworkError>);
}
```